pkg/controller/route53hostedzone/zonereconciler.go

```diff
@@ -0,0 +1,336 @@
+func (zr *ZoneReconciler) getDesiredState() error {
+	if zr.desiredState != nil {
+		// This should never happen.
+		return fmt.Errorf("Desired State already populated for %v/%v", zr.namespace, zr.name)
+	}
+
+	dnszone, err := zr.dnsZonesLister.DNSZones(zr.namespace).Get(zr.name)
+	if err != nil {
+		return err
+	}
+
+	zr.desiredState = dnszone.DeepCopy()
+	zr.dnszone = zr.desiredState.Spec.DNSZone
+
+	return nil
+}
+
+// getCOAWSClient generates a cluster operator aws client.
+// If ZoneReconciler.coAWSClient is not nil, an error will be returned.
+func (zr *ZoneReconciler) getCOAWSClient() error {
+	if zr.coAWSClient != nil {
+		// This should never happen.
+		return fmt.Errorf("coAWSClient already populated for %v/%v", zr.namespace, zr.name)
+	}
+
+	// This allows for using host profiles for AWS auth.
+	secretName := ""
+	regionName := ""
+	if zr.desiredState != nil && zr.desiredState.Spec.AWS != nil {
+		secretName = zr.desiredState.Spec.AWS.AccountSecret.Name
+		regionName = zr.desiredState.Spec.AWS.Region
+	}
+
+	coawsclient, err := zr.funcs.awsClientBuilder(zr.kubeClient, secretName, zr.namespace, regionName)
+	if err != nil {
+		zr.logger.Errorf("Error creating COAWSClient: %v", err)
+		return err
+	}
+
+	zr.coAWSClient = coawsclient
+	return nil
+}
+
+// getCurrentState gets the AWS object for the zone.
+// If ZoneReconciler.currentState is not nil, an error will be returned.
+func (zr *ZoneReconciler) getCurrentState() error {
+	if zr.currentState != nil {
+		// This should never happen.
+		return fmt.Errorf("Current State already populated for %v/%v", zr.namespace, zr.name)
+	}
+
+	output, err := zr.coAWSClient.ListHostedZones(&route53.ListHostedZonesInput{})
+	if err != nil {
+		return err
+	}
+
+	for _, hostedzone := range output.HostedZones {
+		// Find our specific hostedzone
+		cmpStr := zr.dnszone
+
+		if string(cmpStr[len(cmpStr)-1]) != "." {
+			cmpStr = cmpStr + "."
+		}
+
+		if strings.EqualFold(cmpStr, *hostedzone.Name) {
+			zr.currentState = hostedzone
+			return nil
+		}
+	}
+
+	zr.logger.Infof("route53 hostedzone doesn't exist: %v", zr.dnszone)
+
+	// It is ok for i.currentState to still be nil here. That just means it needs to be created.
+	return nil
+}
+
+// Reconcile attempts to make the current state reflect the desired state. It does this idempotently.
+func (zr *ZoneReconciler) Reconcile() error {
+	if zr.desiredState == nil {
+		// There is nothing to reconcile because the desired state isn't present.
+		// This is valid for the case where the controller update event is triggered, which deletes the desired state object and
+		// then a controller delete event is triggered, but the desired state is already gone.
+		return nil
+	}
+
+	// Deletion case
+	if zr.desiredState.DeletionTimestamp != nil {
+		return zr.funcs.deleteRoute53HostedZone()
+	}
+
+	// Creation case
+	if zr.currentState == nil {
+		return zr.funcs.createRoute53HostedZone()
+	}
```

Comment:
Do we have a rate limiting problem like the awselb controller did? Every time this object syncs through we end up hitting the AWS API. It might be good to follow awselb and add a "last sucess" timestamp and only attempt to recreate if it's been over a period of time, or has never been marked successful.